coderd/database/queries/insights.sql

```diff
@@ -771,3 +771,124 @@ SELECT
 FROM unique_template_params utp
 JOIN workspace_build_parameters wbp ON (utp.workspace_build_ids @> ARRAY[wbp.workspace_build_id] AND utp.name = wbp.name)
 GROUP BY utp.num, utp.template_ids, utp.name, utp.type, utp.display_name, utp.description, utp.options, wbp.value;
+
+-- name: GetUserStatusCountsOverTime :many
+-- GetUserStatusCountsOverTime returns the count of users in each status over time.
+-- The time range is inclusively defined by the start_time and end_time parameters.
+--
+-- Bucketing:
+-- Between the start_time and end_time, we include each timestamp where a user's status changed or they were deleted.
+-- We do not bucket these results by day or some other time unit. This is because such bucketing would hide potentially
+-- important patterns. If a user was active for 23 hours and 59 minutes, and then suspended, a daily bucket would hide this.
+-- A daily bucket would also have required us to carefully manage the timezone of the bucket based on the timezone of the user.
+--
+-- Accumulation:
+-- We do not start counting from 0 at the start_time. We check the last status change before the start_time for each user. As such,
+-- the result shows the total number of users in each status on any particular day.
+WITH
+-- dates_of_interest defines all points in time that are relevant to the query.
+-- It includes the start_time, all status changes, all deletions, and the end_time.
+dates_of_interest AS (
+	SELECT @start_time::timestamptz AS date
+
+	UNION
+
+    SELECT DISTINCT changed_at AS date
+    FROM user_status_changes
+    WHERE changed_at > @start_time::timestamptz
+        AND changed_at < @end_time::timestamptz
+
+	UNION
+
+	SELECT DISTINCT deleted_at AS date
+	FROM user_deleted
+	WHERE deleted_at > @start_time::timestamptz
+		AND deleted_at < @end_time::timestamptz
+
+	UNION
+
+	SELECT @end_time::timestamptz AS date
+),
+-- latest_status_before_range defines the status of each user before the start_time.
+-- We do not include users who were deleted before the start_time. We use this to ensure that
+-- we correctly count users prior to the start_time for a complete graph.
```

Comment:
If a user was deleted before the start time, should they not have 0 status changes? A deleted user won't change status going forward right?

Reply:
> We do not include users who were deleted before the start_time. We use this to ensure that
-- we correctly count users prior to the start_time for a complete graph.

There is no issue, mainly clarity. This comment indicated some logic was created to handle cases that should not exist. Now it has been a minute since I looked at this, but it is good :+1: